ext/nnstreamer/tensor_filter/tensor_filter_movidius_ncsdk2.c

```diff
@@ -24,44 +24,248 @@
+  }
 }
 
 /**
  * @brief The open callback for GstTensorFilterFramework. Called before anything else
  * @param prop : property of tensor_filter instance
- * @param private_data : movidius-ncsdk2  plugin's private data
- * @todo : fill this function
+ * @param private_data : movidius-ncsdk2 plugin's private data
+ * @return 0 if OK. -1 if error.
  */
 static int
 _mvncsdk2_open (const GstTensorFilterProperties * prop, void **private_data)
 {
+  /* Variables of the data types from mvnc.h */
+  struct ncDeviceHandle_t *handle_device = NULL;
+  struct ncGraphHandle_t *handle_graph = NULL;
+  struct ncTensorDescriptor_t tensor_desc_input;
+  struct ncTensorDescriptor_t tensor_desc_output;
+  struct ncFifoHandle_t *handle_fifo_input;
+  struct ncFifoHandle_t *handle_fifo_output;
+  ncStatus_t ret_code;
+  /* Normal variables */
+  GMappedFile *file_model = NULL;
+  mvncsdk2_data *pdata = NULL;
+  gint32 sdk_ver[NNS_MVNCSDK2_API_VER_ARRAY_SIZE];
+  guint32 size_sdk_ver = sizeof (sdk_ver);
+  gint32 idx_dev = -1;
+  guint32 len_model_file;
+  void *buf_model_file;
+  guint32 len;
+  gint32 i;
+
+  /* 0. Check the API version */
+  ret_code = ncGlobalGetOption (NC_RO_API_VERSION, sdk_ver, &size_sdk_ver);
+  if ((ret_code != NC_OK) && (sdk_ver[0] != NNS_MVNCSDK2_SUPPORT_API_MAJOR_VER)) {
+    g_printerr
+        ("The major version number of the MVNCSDK API should be %d, not %d\n",
+        NNS_MVNCSDK2_SUPPORT_API_MAJOR_VER, sdk_ver[0]);
+    return -1;
+  }
+
+  /**
+   * 1. Initialize device handle:
+   *  we do not know how many devices are plugged and used. Therefore,
+   *  let's try all the possible device indices (currently,
+   *  0 ~ NNS_MVNCSDK2_SUPPORT_MAX_NUMS_DEVICES) here.
+   */
+  for (i = 0; i < NNS_MVNCSDK2_SUPPORT_MAX_NUMS_DEVICES; ++i) {
+    ret_code = ncDeviceCreate (i, &handle_device);
+    if (ret_code == NC_OK) {
+      idx_dev = i;
+      break;
+    } else {
+      GST_WARNING ("Failed to create device handle at index %d: "
+          "%d is returned\n", i, ret_code);
+    }
+  }
+  if ((ret_code != NC_OK) && (i == NNS_MVNCSDK2_SUPPORT_MAX_NUMS_DEVICES)) {
+    g_printerr ("Cannot create device handle: no available device found\n");
+    return -1;
+  }
+
+  /**
+   * 2. Initialize graph (model) handle
+   */
+  ret_code = ncGraphCreate (prop->model_file, &handle_graph);
+  if (ret_code != NC_OK) {
+    g_printerr ("Cannot create graph handle for \"%s\"\n", prop->model_file);
+    goto err_destroy_device_h;
+  }
+
+  /**
+   * 3. Open device using device handle
+   */
+  ret_code = ncDeviceOpen (handle_device);
+  if (ret_code != NC_OK) {
+    g_printerr ("Cannot open device at index %d\n", idx_dev);
+    goto err_destroy_graph_h;
+  }
+
+  /**
+   * 4. Send model (graph) to the device
+   */
+  file_model = g_mapped_file_new (prop->model_file, FALSE, NULL);
+  if (file_model == NULL) {
+    g_printerr ("Failed to g_mapped_file_new for the model file, \"%s\"\n",
+        prop->model_file);
+    goto err_destroy_graph_h;
+  }
+
```

Comment:
When this `file_model` is assigned to `pdata->gmapped_file_model`?

Reply:
@dongju-chae You are absolutely right. I will fix it.